daml/ContingentClaims/Observation.daml

```diff
@@ -7,140 +7,103 @@
 
 module ContingentClaims.Observation (
     Observation(..)
-  , observe
-  , expiry
   , eval
-  , HasFixings(..)
-  , Key
 ) where
 
 import ContingentClaims.Observable qualified as O
-import Daml.Control.Arrow (Kleisli(..))
-import Daml.Control.Category
+import ContingentClaims.Recursion (cataM)
+import Daml.Control.Arrow (Kleisli(..), arr)
+import Daml.Control.Recursion (Recursive(..), Corecursive(..))
+import DA.Traversable (Traversable(..))
+import DA.Foldable (Foldable(..))
+import Prelude hiding (key, (.))
 
-type Key = Text
 type T = Observation
 
 -- | Concrete implementation of `Observable`, which can be serialized.
-data Observation t a
```

Comment:
I've managed to replace the unholy mess on the left-hand side with a much simpler and cleaner implementation.
In the old version, `a` was a phantom type. Now it's actually a concrete value. Although we always envisaged the use of this DSL inside DAML, The advantage of this is that users can now write much safer `Observation`s off-ledger. Degenerate cases are caught at compile time (if the off-ledger language is strongly typed), or at worst at deserialization time - it never makes it onto the ledger.